src/Payments.sol

```diff
@@ -1754,6 +1783,43 @@ contract Payments is ReentrancyGuard {
 
         return (fundedUntilEpoch, currentFunds, availableFunds, currentLockupRate);
     }
+
+    /**
+     * @notice Burn FIL to buy the network fees
+     * @param token Which kind of fees to buy
+     * @param recipient Receives the purchased fees
+     * @param requested Exact amount of fees transferred
+     */
+    function burnFILForFees(IERC20 token, address recipient, uint256 requested) external payable {
+        Account storage fees = accounts[token][address(this)];
+        uint256 available = fees.funds;
+        require(available >= requested, Errors.WithdrawAmountExceedsAccumulatedFees(token, available, requested));
+
+        AuctionInfo storage auction = auctionInfo[token];
+        uint256 auctionPrice = uint256(auction.startPrice).decay(block.timestamp - auction.startTime);
+        require(msg.value >= auctionPrice, Errors.InsufficientNativeTokenForBurn(msg.value, auctionPrice));
+
+        auction.startPrice = uint88(auctionPrice * Dutch.RESET_FACTOR);
+        auction.startTime = uint168(block.timestamp);
+
+        fees.funds = available - requested;
+
+        (bool success,) = BURN_ADDRESS.call{value: msg.value}("");
+        require(success, Errors.NativeTransferFailed(BURN_ADDRESS, msg.value));
+
+        {
+            uint256 balanceBefore = token.balanceOf(address(this));
+            token.safeTransfer(recipient, requested);
+            uint256 balanceAfter = token.balanceOf(address(this));
+
+            uint256 balanceChange = balanceBefore - balanceAfter;
+            // handle fee-on-transfer and hidden-denominator tokens
+            // re-entry via erc20 transfer can burn unrequested fees, which is fine
```

Comment:
I'm not sure what this comment is trying to say.

In general, one would _usually_ update the funds _first_, then transfer as the very last step to avoid any reentrancy issues. If the transfer fails, you'd simply revert everything.

Reply:
ERC20 tokens can do all sorts of weird things. What this comment is considering is if that ERC20 transfer() gave execution back to a malicious party who can re-enter Payments: what can they do with this power? They can settleRail; they can deposit or withdraw.

> one would usually update the funds first, then transfer as the very last step to avoid any reentrancy issues.

That is what we do, and then we do balanceOf and adjust the auction balance again, but only downward, in case the transfer sent out more tokens than we expected (example of this in `test/mocks/ExtraFeeToken.sol`, tested in `test/BurnExtraFeeToken.t.sol`).

Reply:
Please look at this more carefully. This code is _double accounting_ for deducted funds (funds that were already accounted for on withdrawal).

But as I no longer work here, I can't really spend any more time on this.

Reply:
@zenground0 might have an opinion on this since it's similar to #165. Maybe we also want to do the wrapping `balanceOf` for withdrawal. I should probably also make this method `nonReentrant` to hedge against my own hubris.

Reply:
708662a